tests/test_cases/coreutils_suite/shm_sem_server.c

```diff
@@ -0,0 +1,63 @@
+#include <sys/ipc.h>
+#include <sys/shm.h>
+#include <stdio.h>
+#include <string.h>
+#include <semaphore.h>
+#include <stdlib.h>
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+int main()
+{
+    key_t key = 31337;
+    printf("[server] start...\n");
+    fflush(stdout);
+    // shmget returns an identifier in shmid
+    int shmid = shmget(key, 1024, 0666 | IPC_CREAT);
+
+    // shmat to attach to shared memory
+    sem_t *sem_ptr = (sem_t *)shmat(shmid, NULL, 0);
+
+    key_t key2 = 13371;
+    // shmget returns an identifier in shmid
+    int shmid2 = shmget(key2, 1024, 0666 | IPC_CREAT);
+
+    // shmat to attach to shared memory
+    char *shared_buffer = (char *)shmat(shmid2, NULL, 0);
+
+    // Initialize the semaphore
+    if (sem_init(sem_ptr, 1, 1) < 0)
+    {
+        perror("sem_init");
+        exit(1);
+    }
+
+    int count = 0;
+    while (count++ < 5)
+    {
+        printf("[server] wait\n");
+        fflush(stdout);
+        int status = sem_wait(sem_ptr);
+
+        if (status != 0)
+        {
+            perror("in sem_wait");
+        }
+        sleep(2);
```

Comment:
does this work without these sleeps?

I understand having ones so the right process gets to the semaphors first etc, but once they're both engaged in semaphores the concurrency should be handled by those.

Reply:
There's not really any concurrency without sleeps. It's mostly the first program just keeping the lock and ending and then the next program taking the lock.

Reply:
So I think it's better to keep the locks, I can maybe reduce the amount of sleep